Improve SMB Grunt stability

chopicalqui opened the pull request:
As mentioned in issue #154, connections to SMB Grunts stall unexpectedly. This pull request updates the HTTP and SMB Grunt code to increase pipe communication stability.

**Performed Successful Tests After Code Update**
- Using SMB Grunts for several hours.
- Connect to several SMB Grunts from a single HTTP Grunt and successfully issue commands.
- Connect to a SMB Grunt from another SMB Grunt and successfully issue commands.
- Disconnect and reconnect to a SMB Grunt (from the same Grunt) and successfully issue commands.

**Existing Limitations After Code Update**
- If a SMB communication partner dies, then the communication partner might wait forever to receive data (For intermediate/egress Grunts this is not a big issue as the SMB communication is handled by a separate thread). It seems that an effective fix requires the usage of asynchronous named pipes, which requires rewriting the code base.
- In some cases, output is displayed to end users only after a second command is issued in the UI.
- SMB Grunts might be reported as lost but the are not.
- Creating a second SMB Grunt with the same pipe name in the web interface breaks the SMB Grunt during Stage 0 due to conflicting entries in Covenant's internal database.
- During the disconnect, the string "Disconnect" is sent to the SMB Grunt to kick it out from the read block and to let it return to status "wait for new connection". This hard-coded string poses a network-based indicator, which will be fixed when the SMB Grunt uses asynchronous named pipes (see first point).

**Final Notes**
If you are already working with Covenant and want to use this updated Grunt code, then the recompilation of the Covenant source code is not required.  Instead, you have to perform one of the following tasks:

1. Re-Initialize Covenant database: On the C2, stop Covenant, create a database backup, delete the Covenant database and start Covenant again. **Caution: By performing these steps, you will permanently loose all your existing Grunt connections and you will have to start from scratch.**

```
# a. Stop Covenant
# b. Create database backup
cp Covenant/Covenant/Data/covenant.db Covenant/Covenant/Data/covenant.db.bak
# c. Delete database
rm Covenant/Covenant/Data/covenant.db
# d. Start Covenant
```
2. Manually update Covenant database: On the C2, stop Covenant, create a backup of the Covenant database `covenant.db`, open the Covenant database in your favourite SQLite editor, navigate to table `ImplantTemplates`, and copy the updated HTTP and SMB Grunt source code into column `ExecutorCode` of the respective row. Finally, start Covenant again.

cobbr commented:
Thanks for taking a look at this @chopicalqui! Definitely could use some more stability around SMB grunts.

This one might take me some time to dig into. Once Covenant is updated to v0.5, I'll circle back and give this a thorough review 👍

mofer commented:
Hi @chopicalqui I am following the steps that you described but you say:

> "navigate to table ImplantTemplates, and copy the updated HTTP and SMB Grunt source code into column ExecutorCode of the respective row" 

...what do you mean with "**copy the updated HTTP and SMB Grunt source**" what is the updated source? or you refers to copy the original source code from _Launcher_ > _Binary_ > _Code_ and paste it again in ExecutorCode column? (make no sense to me). 

May you clarify this?

chopicalqui commented:
Hey @mofer

Thanks for your message. The pull request originates from my Github repository and comprises the following two updated files:
- https://github.com/chopicalqui/Covenant/blob/dev/Covenant/Data/Grunt/GruntHTTP/GruntHTTP.cs
- https://github.com/chopicalqui/Covenant/blob/dev/Covenant/Data/Grunt/GruntSMB/GruntSMB.cs

I recommend, verifying any changes that were made in any pull requests you use (see tab `Files Changed` above) - just to be sure nothing malicious was incorporated ;).

If everything looks good and after you created a backup of your Covenant database, you can copy the contents of the above mentioned files into the `ExecutorCode` column of table `ImplantTemplates` (see screenshot below).

![image](https://user-images.githubusercontent.com/39760194/81181130-469c0b80-8fac-11ea-9faa-c2f072c18ca0.png)

Make sure that you copy the content of GruntHTTP.cs into the `ExecutorCode` column of row GruntHTTP and the content of GruntSMB.cs into the `ExecutorCode` column of row GruntSMB.

Afterwards, any newly spawned Grunt will receive the updated compiled assembly code during staging.

Cheers
Chopicalqui

chopicalqui commented:
Hey @cobbr 
I just completed the Red Team Ops course this weekend and I would have time to improve this PR now. Are you working on this issue already or can I support you on this?
Cheers

cobbr commented:
Hey @chopicalqui , apologies I have not had a chance to dig into this PR yet. If you feel there is room for improvement, then definitely feel free to do so!

Flangvik commented:
In regards to the Red Team Ops course, this needs to make it the master repo! Great stuff @chopicalqui and thanks for updating it for Covenant 0.6 @juliourena

chopicalqui commented:
@Flangvik I am currently updating and testing my existing SMB pipe functionality to increase its stability (e.g., allow reconnecting to the SMB grunt even if the peer grunt, which initially connected to the SMB grunt, died). I want to have a thoroughly tested final version before I update the pull request. I am planning to have the updated pull request ready end of August 2020/beginning of September 2020. Cheers

juliourena commented:
That’s some great news @chopicalqui looking forward to it

Flangvik commented:
Great stuff @chopicalqui, can't wait!

sowhatsupeirik commented:
> 
> 
> Great stuff @chopicalqui, can't wait!

Super great! Highly needed change.

chopicalqui commented:
I updated my PR to the latest commit of the upstream dev branch. In addition, I updated the code so that HTTP and SMB Grunts detect broken named pipes. Now, if a named pipe breaks, then the communicating Grunts return back to their initial state (e.g., listening). I added comments to my source code to make my updates/decisions more understandable. This is my final PR version.
While testing my PR, I identified two new issues in regard to SMB Grunts (see links above).
Cheers
Chopicalqui

cobbr commented:
Hey @chopicalqui ! Thanks for all the work on this, and sorry for the late response. Option 2 would definitely be my preference. This _should_ already be possible, if it doesn't then I'd consider it a bug.

chopicalqui commented:
Hey @cobbr No worries. I thought that you'll prefer option 2. Therefore, I filed issue #231 - I wasn't able fixing it as it has something to do with the path calculation.

cobbr commented:
Great! I'll take a look. Going to prioritize this PR and issues for v0.7.

cobbr commented on Covenant/Data/Grunt/GruntSMB/GruntSMB.cs:
```diff
@@ -673,6 +764,68 @@ private byte[] ReadBytes()
+                this.IsConnected = true;
+                this.MonitorPipeState();
+                // Tell the parent Grunt the GUID so that it knows to which child grunt which messages shall be forwarded. Without this message, any further communication breaks.
+                this.Messenger.QueueTaskingMessage(string.Empty);
+                this.Messenger.WriteTaskingMessage();
```
Hey @chopicalqui , could you elaborate a bit on what this is for? It's not clear to me why this is necessary?

chopicalqui replied:
Hey @cobbr 
These two lines are required to successfully reconnect to already **fully staged** SMB Grunts.  They send an empty message to the Egress Grunt, which is then able to extract the SMB Grunt's ID from it and associate the newly established named pipe with the SMB Grunt's ID (see corresponding code of `TaskingMessenger.Connect` below).

```
[...]
ProfileMessage read = downstream.Read();
if (read != null && !string.IsNullOrEmpty(read.Message))
{
    if (string.IsNullOrEmpty(downstream.Identifier))
    {
        GruntEncryptedMessage message = this.Profile.ParseWriteFormat(read.Message);
        if (message.GUID.Length == 20)
        {
            downstream.Identifier = message.GUID.Substring(10);
        }
        else if (message.GUID.Length == 10)
        {
            downstream.Identifier = message.GUID;
        }
    }
    this.UpstreamMessenger.Write(read.Message);
}
[...]
```

Without these two lines, no association between the named pipe and SMB Grunt ID exists, and as a result,
the Egress Grunt does not know where incoming messages from the C2 server should be forwarded (see forwarding code of `TaskingMessenger.ReadTaskingMessage` below) and any further communication breaks.

```
[...]
string json = this.Crafter.Retrieve(gruntMessage);
GruntEncryptedMessage wrappedMessage = GruntEncryptedMessage.FromJson(json);
IMessenger relay = this.DownstreamMessengers.FirstOrDefault(DM => DM.Identifier == wrappedMessage.GUID);
if (relay != null)
{
    relay.Write(this.Profile.FormatReadFormat(wrappedMessage));
}
[...]
```

**Drawback**
Unfortunately, this message is only required by the Egress Grunt. As the Egress Grunt cannot inspect the SMB Grunt's encrypted message, it does not know that it is an empty message and forwards it to the C2 server, which creates unnecessary egress communication.

**Alternative Solution**
The C2 server itself could add the SMB Grunt ID to the `Connect` task. In this case, the Egress Grunt would be able to obtain the SMB Grunt ID directly during the connect. In this case, these two lines wouldn't be required anymore.

cobbr commented:
Thanks for all the hard work @chopicalqui! Glad to finally get this merged. I owe you one! 🍻

Flangvik commented:
Great work @chopicalqui !!

javierantunez commented:
Thank you @chopicalqui and thank you @cobbr Too!
